src/Storages/RabbitMQ/WriteBufferToRabbitMQProducer.cpp

```diff
@@ -52,47 +53,38 @@ WriteBufferToRabbitMQProducer::WriteBufferToRabbitMQProducer(
 
     loop = std::make_unique<uv_loop_t>();
     uv_loop_init(loop.get());
-
     event_handler = std::make_unique<RabbitMQHandler>(loop.get(), log);
-    connection = std::make_unique<AMQP::TcpConnection>(event_handler.get(), AMQP::Address(parsed_address.first, parsed_address.second, AMQP::Login(login_password.first, login_password.second), "/"));
-
-    /* The reason behind making a separate connection for each concurrent producer is explained here:
-     * https://github.com/CopernicaMarketingSoftware/AMQP-CPP/issues/128#issuecomment-300780086 - publishing from
-     * different threads (as outputStreams are asynchronous) with the same connection leads to internal library errors.
-     */
-    size_t cnt_retries = 0;
-    while (!connection->ready() && ++cnt_retries != RETRIES_MAX)
-    {
-        event_handler->iterateLoop();
-        std::this_thread::sleep_for(std::chrono::milliseconds(CONNECT_SLEEP));
-    }
 
-    if (!connection->ready())
-    {
-        throw Exception("Cannot set up connection for producer", ErrorCodes::CANNOT_CONNECT_RABBITMQ);
-    }
+    if (setupConnection(false))
+        setupChannel();
 
-    producer_channel = std::make_shared<AMQP::TcpChannel>(connection.get());
-    checkExchange();
+    writing_task = global_context.getSchedulePool().createTask("RabbitMQWritingTask", [this]{ writingFunc(); });
+    writing_task->deactivate();
 
-    /// If publishing should be wrapped in transactions
-    if (use_transactional_channel)
+    if (exchange_type == AMQP::ExchangeType::headers)
     {
-        producer_channel->startTransaction();
+        for (const auto & header : routing_keys)
+        {
+            std::vector<String> matching;
+            boost::split(matching, header, [](char c){ return c == '='; });
+            key_arguments[matching[0]] = matching[1];
+        }
     }
-
-    writing_task = global_context.getSchedulePool().createTask("RabbitMQWritingTask", [this]{ writingFunc(); });
-    writing_task->deactivate();
 }
 
 
 WriteBufferToRabbitMQProducer::~WriteBufferToRabbitMQProducer()
 {
-    stop_loop.store(true);
     writing_task->deactivate();
-    checkExchange();
-
     connection->close();
+
+    size_t cnt_retries = 0;
+    while (!connection->closed() && ++cnt_retries != (RETRIES_MAX >> 1))
+    {
+        event_handler->iterateLoop();
+        std::this_thread::sleep_for(std::chrono::milliseconds(CONNECT_SLEEP >> 3));
```

Comment:
same here.

Reply:
is it bad to reuse constants in another context but modifying them for needed case? may be I should just change `connect_sleep` to `wait_sleep` not to be case-specific